docs/contributing/data-handling-and-conversion.md

```diff
@@ -0,0 +1,809 @@
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].(bool); ok {
+        apiObject.NestedAttributeName = aws.Bool(v)
+    }
+
+    // ...
+}
+```
+
+To read, if only sending the attribute value when `true` is preferred (`!v` for opposite):
+
+```go
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].(bool); ok && v {
+        apiObject.NestedAttributeName = aws.Bool(v)
+    }
+
+    // ...
+}
+```
+
+To write:
+
+```go
+func flattenServiceStructure(apiObject *service.Structure) map[string]interface{} {
+    // ...
+
+    if v := apiObject.NestedAttributeName; v != nil {
+        tfMap["nested_attribute_name"] = aws.BoolValue(v)
+    }
+
+    // ...
+}
+```
+
+### Nested TypeFloat and AWS Float
+
+To read:
+
+```go
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].(int); ok && v != 0.0 {
+        apiObject.NestedAttributeName = aws.Float64(float64(v))
+    }
+
+    // ...
+}
+```
+
+To write:
+
+```go
+func flattenServiceStructure(apiObject *service.Structure) map[string]interface{} {
+    // ...
+
+    if v := apiObject.NestedAttributeName; v != nil {
+        tfMap["nested_attribute_name"] = aws.Float64Value(v)
+    }
+
+    // ...
+}
+```
+
+### Nested TypeInt and AWS Integer
+
+To read:
+
+```go
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].(int); ok && v != 0 {
+        apiObject.NestedAttributeName = aws.Int64(int64(v))
+    }
+
+    // ...
+}
+```
+
+To write:
+
+```go
+func flattenServiceStructure(apiObject *service.Structure) map[string]interface{} {
+    // ...
+
+    if v := apiObject.NestedAttributeName; v != nil {
+        tfMap["nested_attribute_name"] = aws.Int64Value(v)
+    }
+
+    // ...
+}
+```
+
+### Nested TypeList of Resource and AWS List of Structure
+
+To read:
+
+```go
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].([]interface{}); ok && len(v) > 0 {
+        apiObject.NestedAttributeName = expandServiceStructures(v)
+    }
+
+    // ...
+}
+```
+
+To write:
+
+```go
+func flattenServiceStructure(apiObject *service.Structure) map[string]interface{} {
+    // ...
+
+    if v := apiObject.NestedAttributeName; v != nil {
+        tfMap["nested_attribute_name"] = flattenServiceNestedStructures(v)
+    }
+
+    // ...
+}
+```
+
+### Nested TypeList of Resource and AWS Structure
+
+To read:
+
+```go
+func expandServiceStructure(tfMap map[string]interface{}) *service.Structure {
+    // ...
+
+    if v, ok := tfMap["nested_attribute_name"].([]interface{}); ok && len(v) > 0 {
+        apiObject.NestedAttributeName = expandServiceStructure(v[0].map[string]interface{})
+    }
```

Comment:
There are cases (e.g. #16816 and I've seen others) where a resource is marked as `Required: true` and all its elements are marked `Optional: true`.
In this case a configuration something like

```hcl
resource "aws_svc_thing" "example" {
  flump {}
}
```

indicates that the default value for the Flump property is to be used and this is to be specified via the API as

```go
input.Flump = &svc.Flump{}
```

But `len(v) == 0` so `input.Flump = nil`.

Reply:
I wrote some quick code to log attributes that have this property:

```go
func TestProvider_RequiredContainingAllOptional(t *testing.T) {
	for resourceName, resource := range Provider().ResourcesMap {
		for topLevelAttributeName, v := range resource.Schema {
			if v.Type == schema.TypeList || v.Type == schema.TypeSet {
				switch ty := v.Elem.(type) {
				case *schema.Resource:
					rrrContainer(t, []string{resourceName, topLevelAttributeName}, ty.Schema, false)
				}
			}
		}
	}
}

// The current schema is a container (List or Set)
func rrrContainer(t *testing.T, path []string, m map[string]*schema.Schema, required bool) {
	allOptional := true

	for attributeName, v := range m {
		if v.Type == schema.TypeList || v.Type == schema.TypeSet {
			switch ty := v.Elem.(type) {
			case *schema.Resource:
				rrrContainer(t, append(append([]string{}, path...), attributeName), ty.Schema, v.Required)
			}
		}

		if !(v.Optional && v.Default == nil) {
			allOptional = false
		}
	}

	if required && allOptional {
		t.Logf("%s is Required and has only Optional attributes without Defaults", strings.Join(path, "."))
	}
}
```

```console
$ go test -v ./aws -run TestProvider_RequiredContainingAllOptional
```

and there are only a dozen or so outside the `wafv2` resources.
Some are already covered by having an additional `ExactlyOneOf` attribute.